pkg/plugins/storage/azblob/azblob.go

```diff
@@ -0,0 +1,185 @@
+		context.TODO(),
+		0,
+		azblob.CountToEnd,
+		azblob.BlobAccessConditions{},
+		false,
+		azblob.ClientProvidedKeyOptions{},
+	)
+
+	if err != nil {
+		return nil, newErr(
+			codes.Internal,
+			"Unable to download blob",
+			err,
+		)
+	}
+
+	// TODO: Configure retries
+	data := r.Body(azblob.RetryReaderOptions{MaxRetryRequests: 20})
+
+	return ioutil.ReadAll(data)
+}
+
+func (a *AzblobStorageService) Write(bucket string, key string, object []byte) error {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Write",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+
+	blob := a.getBlobUrl(bucket, key)
+
+	if _, err := blob.Upload(
+		context.TODO(),
+		bytes.NewReader(object),
+		azblob.BlobHTTPHeaders{},
+		azblob.Metadata{},
+		azblob.BlobAccessConditions{},
+		azblob.DefaultAccessTier,
+		nil,
+		azblob.ClientProvidedKeyOptions{},
+	); err != nil {
+		return newErr(
+			codes.Internal,
+			"Unable to write blob data",
+			err,
+		)
+	}
+
+	return nil
+}
+
+func (a *AzblobStorageService) Delete(bucket string, key string) error {
+	newErr := errors.ErrorsWithScope(
+		"AzblobStorageService.Delete",
+		map[string]interface{}{
+			"bucket": bucket,
+			"key":    key,
+		},
+	)
+
+	// Get the bucket for this bucket name
+	blob := a.getBlobUrl(bucket, key)
+
+	if _, err := blob.Delete(
+		context.TODO(),
+		azblob.DeleteSnapshotsOptionInclude,
+		azblob.BlobAccessConditions{},
+	); err != nil {
+		return newErr(
+			codes.Internal,
+			"Unable to delete blob",
+			err,
+		)
+	}
+
+	return nil
+}
+
+const expiryBuffer = 2 * time.Minute
+
+func tokenRefresherFromSpt(spt *adal.ServicePrincipalToken) azblob.TokenRefresher {
+	return func(credential azblob.TokenCredential) time.Duration {
+		if err := spt.Refresh(); err != nil {
+			fmt.Println("Error refreshing token: ", err)
+		} else {
+			tkn := spt.Token()
+			credential.SetToken(tkn.AccessToken)
+
+			return tkn.Expires().Sub(time.Now().Add(expiryBuffer))
+		}
+
+		// Mark the token as already expired
+		return time.Duration(0)
+	}
+}
+
+// New - Creates a new instance of the AzblobStorageService
+func New() (storage.StorageService, error) {
+	// TODO: Create a default storage account for the stack???
+	// XXX: This will limit a membrane wrapped application
+	// to accessing a single storage account
+	storageAccountName := utils.GetEnv(azureutils.AZURE_STORAGE_ACCOUNT_NAME_ENV, "")
+	if storageAccountName == "" {
+		return nil, fmt.Errorf("failed to determine Azure Storage Account Name, environment variable %s not set", azureutils.AZURE_STORAGE_ACCOUNT_NAME_ENV)
+	}
+
+	spt, err := azureutils.GetServicePrincipalToken(azure.PublicCloud.ResourceIdentifiers.Storage)
+	if err != nil {
+		return nil, err
+	}
+
+	cTkn := azblob.NewTokenCredential(spt.Token().AccessToken, tokenRefresherFromSpt(spt))
+
+	var accountURL *url.URL
+	if accountURL, err = url.Parse(fmt.Sprintf("https://%s.blob.core.windows.net", storageAccountName)); err != nil {
```

Comment:
What if the user has a custom domain ? 
Would it be better to have "https://%s.blob.core.windows.net" exposed as an ENV configuration?

Reply:
done, set for both queue and blob, and adding as part of env variables for app service deployment.